Feature/update change password error message

surajeaton opened the pull request:
<!-- If this pull request fixes an Issue, link it below. If not, you can remove the line below -->

Fixes # BLUI-3944

<!-- Include a bulleted list summarizing the main changes you have made in this PR -->

#### Changes proposed in this Pull Request:

- Added confirm password field error state
- Updated test case

<!-- Include screenshots if they will help illustrate the changes in this PR -->

#### Screenshots / Screen Recording (if applicable)

https://user-images.githubusercontent.com/120575281/231490131-29bd2a0d-b69a-4d94-b674-dd044883dad4.mov

<!-- Instruction for PR reviewers, if more complicated than a simple yarn start -->

#### To Test:

- yarn start:example

<!-- Useful for draft pull requests -->

#### Any specific feedback you are looking for?

- Please open this PR, once https://github.com/etn-ccis/blui-react-auth-shared/pull/103 gets merged

huayunh commented:
From the story description,

> when the passwords do match, display a green checkmark / tick in the confirm password field.

Are you planning to do it this PR, or in a separate PR?

huayunh commented:
I also think someone may have changed the design intention in the story description. Here's what I expect to happen — it will be a mix of onChange and onBlur event listener:

- The user types their new password
- All requirements are met
- The user focuses on the "Confirm Password" field
  - No error text at this stage
- The user keeps typing
  - No error text at this stage
  - While typing, if password matches, display the green checkmark. So this check mark check is done "onChange". But do not display "passwords do not match" just yet
- The user blurs away from the "Confirm Password" field
  - If not matching, show the error state and the error text
  - if matching, do nothing (keep displaying the green checkmark).
  - Submit button stays disabled
- The user re-focus on the "Confirm Password" text field, hoping to correct their error.
  - the error state & error text persists until the user types in matching passwords
    - Even when they are changing the password in the "New Password" field, the error state of the "Confirm Password" field should still display until the two passwords match
  - The error state goes away as soon as the two passwords match, and the green checkmark displays.

The core idea here is to prevent the software from "rushing" the user while users are still typing, and the software should only show the error state when users finish typing. And show a green checkmark when password matches.

surajeaton commented:
> I also think someone may have changed the design intention in the story description. Here's what I expect to happen — it will be a mix of onChange and onBlur event listener:
> 
> * The user types their new password
> * All requirements are met
> * The user focuses on the "Confirm Password" field
>   
>   * No error text at this stage
> * The user keeps typing
>   
>   * No error text at this stage
>   * While typing, if password matches, display the green checkmark. So this check mark check is done "onChange". But do not display "passwords do not match" just yet
> * The user blurs away from the "Confirm Password" field
>   
>   * If not matching, show the error state and the error text
>   * if matching, do nothing (keep displaying the green checkmark).
>   * Submit button stays disabled
> * The user re-focus on the "Confirm Password" text field, hoping to correct their error.
>   
>   * the error state & error text persists until the user types in matching passwords
>     
>     * Even when they are changing the password in the "New Password" field, the error state of the "Confirm Password" field should still display until the two passwords match
>   * The error state goes away as soon as the two passwords match, and the green checkmark displays.
> 
> The core idea here is to prevent the software from "rushing" the user while users are still typing, and the software should only show the error state when users finish typing. And always show a green checkmark when the password matches.
> 
> This behavior on the "Confirm Password" field is expected across all three places where a user needs to confirm their newly-created password — registration, changing the password while logged in, and resetting the password as a result of forgetting the original password.

@huayunh Addressed the feedback. Please review.

huayunh commented:
For an invalid "New Password", the "passwords do not match" still gets triggered even when the passwords do match.

![image](https://user-images.githubusercontent.com/8997218/231785279-fca42a33-fc97-44ff-8317-51f997d8a662.png)

Can you do something to make it behave like this: 
- onBlur, check the New Password field
  - if invalid, put it in the error state
    - when users continue to make changes, do the check onChange
    - This logic will be similar to the "Passwords do not match" — display error message onBlur, and keep doing the checks onChange until the user finally gets it

![image](https://user-images.githubusercontent.com/8997218/231789956-7980c018-5811-46d5-bbea-c8f47e5ead19.png)

surajeaton commented:
> For an invalid "New Password", the "passwords do not match" still gets triggered even when the passwords do match.
> 
> ![image](https://user-images.githubusercontent.com/8997218/231785279-fca42a33-fc97-44ff-8317-51f997d8a662.png)
> 
> Can you do something to make it behave like this:
> 
> * onBlur, check the New Password field
>   
>   * if invalid, put it in the error state
>     
>     * when users continue to make changes, do the check onChange
>     * This logic will be similar to the "Passwords do not match" — display error message onBlur, and keep doing the checks onChange until the user finally gets it
> 
> ![image](https://user-images.githubusercontent.com/8997218/231789956-7980c018-5811-46d5-bbea-c8f47e5ead19.png)

@huayunh  Done. Please review

huayunh commented:
Code logic works well, but saw this in the console

<img width="1433" alt="image" src="https://user-images.githubusercontent.com/8997218/232069027-52be91bc-588e-48e9-a26d-3a75b2a173e2.png">
